### PR TITLE
feat(restaurant-showcase): add miranum showcase using kafka

### LINKED ISSUE #3: [Epic] - Evolve Pizza-Order-Showcase
- [x] Description created
- [x] Reviewed by peer: @DaAnda97 
- [ ] Prioritised

# Description
This epic sums up all tasks which needs to be done to deliver a talk at the Camunda Community Summit. During this effort the pizza-order-showcase needs to be changed in such a way that it can be used for a talk.  

## Sub-Tasks
 
```[tasklist]
### Tasks
- [ ] #4 
- [ ] #5
- [ ] #6 
- [ ] #7
- [ ] #8 
```

## Stakeholder
* @DaAnda97 
* @Hafflgav 

# Acceptance
When a talk with a streamlined story can be created out of the showcase.

### LINKED ISSUE #4: [Feature] -
- [x] Description created
- [ ] Reviewed by peer: @DaAnda97 
- [ ] Prioritised

# Description
Migrate[ pizza-e-vino example](https://github.com/Hafflgav/pizza-e-vino) to the pizza-order-showcase. So we can build a complete example in this mono-repo

## Sub-Tasks
 
```[tasklist]
### Tasks
- [ ] Copy the pizza-e-vino project to the pizza-order-showcase
- [ ] Update the package structure
- [ ] Update the readme's 
- [ ] Update the references in the documentation 
```

### LINKED ISSUE #5: [Feature] - Update Process Model for Domain Events
- [x] Description created
- [ ] Reviewed by peer: @DaAnda97 
- [ ] Prioritised

# Description
Change the process model in such a way that Domain Events can be thrown in a meaningful way. 
This can be by adding another little process which is contextually separated or by dealing with intermediate message events. 

## Sub-Tasks
 
```[tasklist]
### Tasks
- [ ] Brainstorm possible ideas and discuss them 
- [ ] Change the process model and ensure that it still can run 
- [ ] Implement new services if necessary 
```

### LINKED ISSUE #6: [Feature] - Unify the Pizza-order-showcase
- [x] Description created
- [x] Reviewed by peer: @DaAnda97 
- [ ] Prioritised

# Description
Unify the pizza-order-showcase so it builds on top of the pizza-e-vino example. Make sure they are named equally and that it is clear that the pizza-order-showcase with the clean architecture is just another iteration of the same process. 

## Sub-Tasks
 
```[tasklist]
### Tasks
- [ ] Update all BPMN files accordingly 
- [ ] Improve naming where necessary. 
```

### LINKED ISSUE #7: [Feature] - Add Kafka and Domain Events to Pizza-Order-Showcase
- [x] Description created
- [x] Reviewed by peer: @DaAnda97 
- [x] Prioritised

# Description
To demonstrate a loosely coupled architecture which follows the guidelines of domain driven design we want to throw domain events and publish them to Kafka. 

## Sub-Tasks
```[tasklist]
### Tasks
- [x] Setup and configure Kafka. Add it to the Stack
- [x] Add code to throw domain events
```

### LINKED ISSUE #8: [Feature] - Implement a proxy to correlate events from Kafka to Camunda
- [x] Description created
- [x] Reviewed by peer: @DaAnda97 
- [ ] Prioritised

# Description
Implement a proxy which correlates events from Kafka with Camunda to show how domain events can trigger other processes in different domains or resume an asynchronous process. Ideally a hexagonal architecture can be used for this. 

## Sub-Tasks
```[tasklist]
### Tasks
- [ ] Implement Proxy which consumes events from Kafka and throws them to Camunda 
```

### PR DESCRIPTION
**Pull request by DaAnda97**
**Issue**
closes #3
closes #4
closes #5
closes #6
closes #7
closes #8

**Description**
Good and Bad Practice Example using Camunda 8

**Checklist**
* [x] Code is easy to understand
* [x] No obvious errors or bugs
* [x] CI/CD Pipelines did run successfully
* [x] Tests were implemented
* [x] Documentation was created

### CODE REVIEW

**Review by Hafflgav — APPROVED** [review]
Runs on my machine

File: pizza-order-showcase/waiter/waiter-core/pom.xml [hunk 1]
```diff
@@ -32,7 +32,7 @@
         <dependency>
             <groupId>io.miragon.miranum</groupId>
             <artifactId>process-api</artifactId>
-            <version>0.1.0-SNAPSHOT</version>
+            <version>${project.version}</version>
```
Comment (rami-nk) [inline pizza-order-showcase/waiter/waiter-core/pom.xml:35]: dfgdfg

File: restaurant-showcase/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+![Restaurant UI](./img/restaurant-user-interface.png)
+
+## 1. Bad Example
+In a first approach the waiter team designed a bpmn process and uses Camunda 8 Rest Connectors to call the services.
+![Restaurant Plain](./img/restaurant-plain-bpmn.png)
+
+### 1.1 Run
+Follow these steps to run the example:
+1. Start the docker services:
```
Comment (rami-nk) [inline restaurant-showcase/README.md:15]: to start Camunda 8 Platform

File: restaurant-showcase/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    ```
+    docker-compose -f stack/docker-compose.yml up -d
+    ```
+2. Deploy the bpmn process to Camunda 8:
+    Open Camunda Modeller and deploy the process model `waiter-plain/src/main/resources/restaurant-plain.bpmn`
```
Comment (rami-nk) [inline restaurant-showcase/README.md:20]: autodeploy

File: restaurant-showcase/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+3. Start the applications: `waiter-plain`, `restaurant` and `kitchen`
+
```
Comment (rami-nk) [inline restaurant-showcase/README.md:22]: Add ports where services can be found

File: restaurant-showcase/README.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+Together with the Miragon consultants, the waiter team designed a new bpmn process, to address those issues.
+The new architecture is based on the following principles:
+- Avoid technical aspects in bmpn models by using External Tasks.
+- Usage of Domain Driven Design and a Hexagonal Architecture to implement adapters for external services. That makes it easy to switch from one provider to another without changing th domain.
+- Avoid boilerplate code by using Miranum Connect to communicate with Camunda.
+
+![Restaurant Plain](./img/restaurant-miranum-bpmn.png)
+
+### 2.1 Run
+Shut down `waiter-plain` and start `waiter-miranum`.
```
Comment (rami-nk) [inline restaurant-showcase/README.md:39]: "To make it run, just shutdown the waiter-plain and start the waiter-miranum."

File: restaurant-showcase/payment/src/main/java/restaurant/showcase/payment/PaymentController.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+@Slf4j
+@RestController
+public class PaymentController {
+
+    private final static long ONE_SECOND = 1000L;
+
+    @RequestMapping(value = "/payment", method = RequestMethod.POST)
+    public String payment(@RequestBody OrderRO orderRO) throws InterruptedException {
+        for (int i = 0; i < 5; i++) {
```
Comment (rami-nk) [inline restaurant-showcase/payment/src/main/java/restaurant/showcase/payment/PaymentController.java:17]: Magic number 5

File: restaurant-showcase/satisfaction-miranum/pom.xml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+            <artifactId>spring-kafka</artifactId>
+        </dependency>
+
+        <!--Miranum-->
+        <dependency>
+            <groupId>io.miragon.miranum</groupId>
+            <artifactId>message-api</artifactId>
+            <version>${miranum.version}</version>
+        </dependency>
+        <dependency>
```
Comment (rami-nk) [inline restaurant-showcase/satisfaction-miranum/pom.xml:36]: we always separated core and the starter impl, why didn't we do that here?

File: restaurant-showcase/satisfaction-miranum/src/main/java/restaurant/showcase/satisfaction/adapter/in/kafka/KafkaConsumer.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+@Log4j2
+@Component
+@RequiredArgsConstructor
+public class KafkaConsumer {
+    private final StartSatisfactionProcessUseCase startSatisfactionProcessUseCase;
+
+    @Value("${kafka.in.restaurant.process.start}")
+    private String processStartEvent;
+
+    @KafkaListener(topics = "#{'${kafka.in.restaurant.topics}'.replaceAll('\\s+', '').split(',')}")
```
[the remark this answers is not in the document]
Comment (rami-nk) [reply restaurant-showcase/satisfaction-miranum/src/main/java/restaurant/showcase/satisfaction/adapter/in/kafka/KafkaConsumer.java:22]: Can this be extracted into a properties wrapper that does this stuff?

File: restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/WaiterMiranumApplication.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package restaurant.showcase.waiter;
+
+import io.camunda.zeebe.spring.client.EnableZeebeClient;
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+@EnableZeebeClient
```
Comment (rami-nk) [inline restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/WaiterMiranumApplication.java:8]: deprecated

File: restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/adapter/in/rest/WaiterController.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+import restaurant.showcase.waiter.adapter.in.rest.dtos.PayMealRequestDto;
+import restaurant.showcase.waiter.adapter.in.rest.dtos.PlaceOrderRequestDto;
+import restaurant.showcase.waiter.adapter.in.rest.dtos.PlaceOrderResponseDto;
+import restaurant.showcase.waiter.application.port.in.payMeal.PayMealInCommand;
+import restaurant.showcase.waiter.application.port.in.payMeal.PayMealUseCase;
+import restaurant.showcase.waiter.application.port.in.placeOrder.PlaceOrderInCommand;
+import restaurant.showcase.waiter.application.port.in.placeOrder.PlaceOrderUseCase;
+
+@RestController
+@Log4j2
```
Comment (rami-nk) [inline restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/adapter/in/rest/WaiterController.java:20]: Slf4j instad

File: restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/adapter/out/kafka/KafkaAdapter.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+import lombok.AllArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.springframework.context.event.EventListener;
+import org.springframework.kafka.core.KafkaTemplate;
+import restaurant.showcase.waiter.domain.events.MealPayedEvent;
+import restaurant.showcase.waiter.domain.events.order.OrderDomainEvent;
+
+import java.util.HashMap;
+
+@Log4j2
```
Comment (rami-nk) [inline restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/adapter/out/kafka/KafkaAdapter.java:12]: Slf4j instead

File: restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/application/port/out/issueCheck/IssueCheckOutCommand.java [hunk 12]
```diff
@@ -0,0 +1,21 @@
+package restaurant.showcase.waiter.application.port.out.issueCheck;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import lombok.RequiredArgsConstructor;
+import restaurant.showcase.waiter.domain.Order;
+
+@RequiredArgsConstructor
+@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
```
Comment (rami-nk) [inline restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/application/port/out/issueCheck/IssueCheckOutCommand.java:8]: Why is this necessary?

File: restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/application/port/out/payMeal/PayMealOutCommand.java [hunk 13]
```diff
@@ -0,0 +1,12 @@
+package restaurant.showcase.waiter.application.port.out.payMeal;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+
```
Comment (rami-nk) [inline restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/application/port/out/payMeal/PayMealOutCommand.java:6]: blank line

File: restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/domain/Food.java [hunk 14]
```diff
@@ -0,0 +1,24 @@
+package restaurant.showcase.waiter.domain;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+import java.util.Arrays;
+
+@RequiredArgsConstructor
+@Getter
+public enum Food {
```
Comment (rami-nk) [inline restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/domain/Food.java:10]: blank line

File: restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/domain/Food.java [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+    HAMBURGER("hamburger"),
+    CHEESEBURGER("cheeseburger"),
+    PIZZA_MARGARITA("pizza-margarita"),
+    PIZZA_PROSCIUTTO("pizza-prosciutto");
+
```
Comment (rami-nk) [inline restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/domain/Food.java:15]: why not just use enum name?

File: restaurant-showcase/waiter-plain/src/main/java/restaurant/showcase/waiter/WaiterPlainApplication.java [hunk 16]
```diff
@@ -0,0 +1,13 @@
+package restaurant.showcase.waiter;
+
+import io.camunda.zeebe.spring.client.EnableZeebeClient;
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+@EnableZeebeClient
```
Comment (rami-nk) [inline restaurant-showcase/waiter-plain/src/main/java/restaurant/showcase/waiter/WaiterPlainApplication.java:8]: deprecated

File: pom.xml [hunk 17]
```diff
@@ -12,8 +12,10 @@
 
     <modules>
         <module>pizza-order-showcase</module>
+        <module>restaurant-showcase</module>
         <module>process-example</module>
         <module>schema-client-example</module>
         <module>customer-onboarding-showcase</module>
+        <module>restaurant-showcase/satisfaction-miranum</module>
```
Comment (rami-nk) [inline pom.xml:19]: Why this?

File: pizza-order-showcase/waiter/waiter-core/pom.xml [hunk 18]
```diff
@@ -32,7 +32,7 @@
...
             <scope>compile</scope>
```
Comment (rami-nk) [inline pizza-order-showcase/waiter/waiter-core/pom.xml:36]: compile is default

File: restaurant-showcase/README.md [hunk 19]
```diff
@@ -0,0 +1,45 @@
+# Restaurant Showcase
```
[the remark this answers is not in the document]
Comment (rami-nk) [reply restaurant-showcase/README.md:1]: And we need guidelines on how to format the README. 
- Should images be centered or not?
- Do we want any image height restrictions do avoid inconsistent and weird looking READMEs?

File: restaurant-showcase/satisfaction-miranum/src/main/java/restaurant/showcase/satisfaction/adapter/in/kafka/KafkaConsumer.java [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+import lombok.extern.log4j.Log4j2;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.kafka.annotation.KafkaListener;
+import org.springframework.stereotype.Component;
+import restaurant.showcase.satisfaction.application.port.in.startSatisfaction.StartSatisfactionProcessInCommand;
+import restaurant.showcase.satisfaction.application.port.in.startSatisfaction.StartSatisfactionProcessUseCase;
+
+import java.util.HashMap;
+
+@Log4j2
```
[the remark this answers is not in the document]
Comment (DaAnda97) [reply restaurant-showcase/satisfaction-miranum/src/main/java/restaurant/showcase/satisfaction/adapter/in/kafka/KafkaConsumer.java:13]: You're right. This needs to be changed.

File: restaurant-showcase/kitchen/src/main/java/restaurant/showcase/kitchen/KitchenController.java [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+@Slf4j
+@RestController
+public class KitchenController {
+
+    private final static long ONE_SECOND = 1000L;
+
+    @RequestMapping(value = "/order", method = RequestMethod.POST)
+    public String order(@RequestBody OrderRO orderRO) throws InterruptedException {
```
[the remark this answers is not in the document]
Comment (DaAnda97) [reply restaurant-showcase/kitchen/src/main/java/restaurant/showcase/kitchen/KitchenController.java:16]: Good Point

File: restaurant-showcase/payment/src/main/java/restaurant/showcase/payment/PaymentController.java (hunk 6 above)
Comment (DaAnda97) [reply restaurant-showcase/payment/src/main/java/restaurant/showcase/payment/PaymentController.java:17]: yes. it's random, since it's just logging.. would let it as is

File: restaurant-showcase/satisfaction-miranum/src/main/java/restaurant/showcase/satisfaction/adapter/in/kafka/KafkaConsumer.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (DaAnda97) [reply restaurant-showcase/satisfaction-miranum/src/main/java/restaurant/showcase/satisfaction/adapter/in/kafka/KafkaConsumer.java:22]: graps all kafka topics configured in applicaition.yaml notated in rows. Could add a comment here

File: restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/application/port/out/issueCheck/IssueCheckOutCommand.java (hunk 12 above)
Comment (DaAnda97) [reply restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/application/port/out/issueCheck/IssueCheckOutCommand.java:8]: Forgot to delete that after refactoring.

File: restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/domain/Food.java (hunk 15 above)
Comment (DaAnda97) [reply restaurant-showcase/waiter-miranum/src/main/java/restaurant.showcase.waiter/domain/Food.java:15]: That's how I'm doing input validation.

File: pom.xml (hunk 17 above)
Comment (DaAnda97) [reply pom.xml:19]: maven placed it there after creating a new module. I think that can be deleted

**Comment by DaAnda97** [comment]
Thanks for your review @rami-nk
